client-sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -0,0 +1,249 @@
+                        sniffedNodes.remove(excludeHost);
+                    }
+                    logger.debug("sniffed nodes: " + sniffedNodes);
+                    this.restClient.setHosts(sniffedNodes.toArray(new HttpHost[sniffedNodes.size()]));
+                } catch (Throwable t) {
+                    logger.error("error while sniffing nodes", t);
+                } finally {
+                    try {
+                        //regardless of whether and when the next sniff is scheduled, cancel it and schedule a new one with updated delay
+                        this.scheduledFuture.cancel(false);
+                        logger.debug("scheduling next sniff in " + nextSniffDelay + " ms");
+                        this.scheduledFuture = this.scheduledExecutorService.schedule(this, nextSniffDelay, TimeUnit.MILLISECONDS);
+                    } catch (Throwable t) {
+                        logger.error("error while scheduling next sniffer task", t);
+                    } finally {
+                        this.nextSniffDelay = sniffInterval;
+                        running.set(false);
+                    }
+                }
+            }
+        }
+
+        void shutdown() {
+            scheduledExecutorService.shutdown();
+            try {
+                if (scheduledExecutorService.awaitTermination(1000, TimeUnit.MILLISECONDS)) {
+                    return;
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            scheduledExecutorService.shutdownNow();
+        }
+    }
+
+    /**
+     * Returns a new {@link Builder} to help with {@link Sniffer} creation.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Sniffer builder. Helps creating a new {@link Sniffer}.
+     */
+    public static final class Builder {
+        public static final long DEFAULT_SNIFF_INTERVAL = TimeUnit.MINUTES.toMillis(5);
+        public static final long DEFAULT_SNIFF_AFTER_FAILURE_DELAY = TimeUnit.MINUTES.toMillis(1);
+        public static final long DEFAULT_SNIFF_REQUEST_TIMEOUT = TimeUnit.SECONDS.toMillis(1);
+
+        private long sniffRequestTimeout = DEFAULT_SNIFF_REQUEST_TIMEOUT;
+        private long sniffInterval = DEFAULT_SNIFF_INTERVAL;
+        private boolean sniffOnFailure = true;
+        private long sniffAfterFailureDelay = DEFAULT_SNIFF_AFTER_FAILURE_DELAY;
+        private String scheme = "http";
+        private RestClient restClient;
+
+        private Builder() {
+
+        }
+
+        /**
+         * Sets the interval between consecutive ordinary sniff executions. Will be honoured when sniffOnFailure is disabled or
+         * when there are no failures between consecutive sniff executions.
+         * @throws IllegalArgumentException if sniffInterval is not greater than 0
+         */
+        public Builder setSniffInterval(int sniffInterval) {
```

Comment:
or just use millis everywhere for now? I am renaming all members and arguments to include their time unit in their name for now.